test/performance/odp_lock_perf.c

```diff
@@ -0,0 +1,670 @@
+/* Copyright (c) 2021, Nokia
+ *
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+#include <inttypes.h>
+#include <stdlib.h>
+#include <getopt.h>
+
+#include <odp_api.h>
+#include <odp/helper/odph_api.h>
+
+/* Max number of workers if num_cpu=0 */
+#define DEFAULT_MAX_WORKERS 10
+
+/* Max number of counters */
+#define MAX_COUNTERS 8
+
+#define TEST_INFO(name, test, validate) { name, test, validate }
+
+typedef enum place_t {
+	PLACE_PACK,
+	PLACE_SEPARATE,
+	PLACE_ALL_SEPARATE,
+} place_t;
+
+/* Command line options */
+typedef struct test_options_t {
+	uint32_t num_cpu;
+	uint32_t type;
+	uint32_t num_round;
+	uint32_t num_counter;
+	place_t place;
+} test_options_t;
+
+/* command line options default values */
+static test_options_t test_options_def = {
+	.num_cpu = 0,
+	.type = 0,
+	.num_round = 100000,
+	.num_counter = 2,
+	.place = 2,
+};
+
+typedef struct test_global_t test_global_t;
+
+/* Test function template */
+typedef void (*test_fn_t)(test_global_t *g, volatile uint64_t **counter,
+			  uint32_t num_counter);
+/* Test result validation function template */
+typedef int (*validate_fn_t)(test_global_t *g, volatile uint64_t **counter,
+			     uint32_t num_counter);
+
+/* Worker thread context */
+typedef struct test_thread_ctx_t {
+	test_global_t *global;
+	test_fn_t func;
+	uint64_t nsec;
+	uint32_t idx;
+} test_thread_ctx_t;
+
+/* Global data */
+struct test_global_t {
+	test_options_t test_options;
+	uint32_t cur_type;
+	odp_barrier_t barrier;
+	odp_cpumask_t cpumask;
+	odph_thread_t thread_tbl[ODP_THREAD_COUNT_MAX];
+	test_thread_ctx_t thread_ctx[ODP_THREAD_COUNT_MAX];
+	struct {
+		struct ODP_ALIGNED_CACHE {
+			odp_spinlock_t lock;
+			volatile uint64_t counter[MAX_COUNTERS];
+		} spinlock;
+		struct ODP_ALIGNED_CACHE {
+			odp_spinlock_recursive_t lock;
+			volatile uint64_t counter[MAX_COUNTERS];
+		} spinlock_recursive;
+		struct ODP_ALIGNED_CACHE {
+			odp_rwlock_t lock;
+			volatile uint64_t counter[MAX_COUNTERS];
+		} rwlock;
+		struct ODP_ALIGNED_CACHE {
+			odp_rwlock_recursive_t lock;
+			volatile uint64_t counter[MAX_COUNTERS];
+		} rwlock_recursive;
+		struct ODP_ALIGNED_CACHE {
+			odp_ticketlock_t lock;
+			volatile uint64_t counter[MAX_COUNTERS];
+		} ticketlock;
+		struct ODP_ALIGNED_CACHE {
+			volatile uint64_t counter[MAX_COUNTERS];
+		} separate;
+		struct {
+			volatile uint64_t ODP_ALIGNED_CACHE counter;
+		} all_separate[MAX_COUNTERS];
+	} item;
+};
+
+typedef struct {
+	const char *name;
+	test_fn_t test_fn;
+	validate_fn_t validate_fn;
+} test_case_t;
+
+static test_global_t *test_global;
+
+static inline void test_spinlock(test_global_t *g, volatile uint64_t **counter,
+				 uint32_t num_counter)
+{
+	odp_spinlock_t *lock = &g->item.spinlock.lock;
+
+	for (uint64_t i = 0; i < (uint64_t)g->test_options.num_round; i++) {
```

Comment:
num_round option could be uint64_t. So that test run time is not limited to 4 billion rounds.